File: src/commands/add.ts
import * as p from '@clack/prompts';
import chalk from 'chalk';
import { execa } from 'execa';
import { join } from 'node:path';
import { exists } from '../utils/fs.js';
import { isValidAddon, listAvailableAddons, getAddonManifest, installAddon } from '../addons/index.js';

export async function add(addon: string): Promise<void> {
	const cwd = process.cwd();

	p.intro(chalk.bold(`forge add ${addon}`));

	// Validate addon name
	if (!isValidAddon(addon)) {
		p.log.error(`Unknown addon: "${addon}"`);
		p.log.message(`Available addons: ${listAvailableAddons().join(', ')}`);
		process.exit(1);
	}

	// Check forge.yaml exists
	if (!(await exists(join(cwd, 'forge.yaml')))) {
		p.log.error('No forge.yaml found. Run `forge init` first.');
		process.exit(1);
	}

	// Check if already installed
	const { readYaml } = await import('../utils/yaml.js');
	const config = await readYaml<Record<string, unknown>>(join(cwd, 'forge.yaml'));
	const addons = (config.addons as string[]) ?? [];
	if (addons.includes(addon)) {
		p.log.warn(`Addon "${addon}" is already installed.`);
		process.exit(0);
	}

	const spinner = p.spinner();
	spinner.start(`Installing ${addon}...`);

	try {
		const installedFiles = await installAddon(addon, cwd);

		for (const file of installedFiles) {
			p.log.success(`Created ${file}`);
		}
		p.log.success(`Updated forge.yaml`);

		// Run post-install commands
		const manifest = await getAddonManifest(addon);
		if (manifest.post_install) {
			for (const cmd of manifest.post_install) {
				spinner.message(`Running: ${cmd}`);
				try {
					await execa(cmd, { shell: true, cwd, stdio: 'pipe' });
					p.log.success(`Ran: ${cmd}`);
				} catch {
					p.log.warn(`Post-install command failed: ${cmd}`);
					p.log.message('  You may need to run this manually.');
				}
			}
		}

		spinner.stop(`${addon} installed`);
	} catch (err) {
		spinner.stop('Installation failed');
		p.log.error(String(err));
		process.exit(1);
	}

	p.outro(chalk.green('Done!'));
}

File: cmd/add.go
package cmd

import (
	"github.com/samahlstrom/forge-cli/internal/addon"
	"github.com/samahlstrom/forge-cli/internal/ui"
	"github.com/samahlstrom/forge-cli/internal/util"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/spf13/cobra"
)

func init() {
	rootCmd.AddCommand(&cobra.Command{
		Use:   "add <addon>",
		Short: "Install an optional addon (e.g., browser-testing, compliance-hipaa)",
		Args:  cobra.ExactArgs(1),
		RunE:  runAdd,
	})
}

func runAdd(cmd *cobra.Command, args []string) error {
	name := args[0]
	cwd, _ := os.Getwd()

	ui.Intro(ui.Bold(fmt.Sprintf("forge add %s", name)))

	if !addon.IsValid(name) {
		ui.Log.Error(fmt.Sprintf("Unknown addon: %q", name))
		ui.Log.Message(fmt.Sprintf("Available addons: %s", strings.Join(addon.ListAvailable(), ", ")))
		os.Exit(1)
	}

	if !util.Exists(filepath.Join(cwd, "forge.yaml")) {
		ui.Log.Error("No forge.yaml found. Run `forge init` first.")
		os.Exit(1)
	}

	var config map[string]any
	_ = util.ReadYAML(filepath.Join(cwd, "forge.yaml"), &config)
	addons := toStringSlice(config["addons"])
	for _, a := range addons {
		if a == name {
			ui.Log.Warn(fmt.Sprintf("Addon %q is already installed.", name))
			os.Exit(0)
		}
	}

	spinner := ui.NewSpinner()
	spinner.Start(fmt.Sprintf("Installing %s...", name))

	files, err := addon.Install(name, cwd)
	if err != nil {
		spinner.Stop("Installation failed")
		ui.Log.Error(err.Error())
		os.Exit(1)
	}

	for _, f := range files {
		ui.Log.Success(fmt.Sprintf("Created %s", f))
	}
	ui.Log.Success("Updated forge.yaml")

	// Post-install commands
	manifest, _ := addon.GetManifest(name)
	for _, cmdStr := range manifest.PostInstall {
		spinner.SetMessage(fmt.Sprintf("Running: %s", cmdStr))
		c := exec.Command("bash", "-c", cmdStr)
		c.Dir = cwd
		if err := c.Run(); err != nil {
			ui.Log.Warn(fmt.Sprintf("Post-install command failed: %s", cmdStr))
			ui.Log.Message("  You may need to run this manually.")
		} else {
			ui.Log.Success(fmt.Sprintf("Ran: %s", cmdStr))
		}
	}

	spinner.Stop(fmt.Sprintf("%s installed", name))
	ui.Outro(ui.Green("Done!"))
	return nil
}
